cli/internal/publish/images.go

```diff
@@ -0,0 +1,367 @@
+	})
+
+	// replace in reverse order
+	for _, ir := range reps {
+		literalTag := contents[ir.startPos:ir.endPos]
+		replacedTag := strings.Replace(literalTag, ir.ref, ir.url, 1)
+		contents = contents[:ir.startPos] + replacedTag + contents[ir.endPos:]
+	}
+
+	return contents, nil
+}
+
+func uploadImage(ctx context.Context, uploadURL, file string) error {
+	f, err := os.Open(file)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	first512 := make([]byte, 512)
+	if _, err := f.Read(first512); err != nil {
+		return err
+	}
+	ct := http.DetectContentType(first512)
+	if _, err := f.Seek(0, io.SeekStart); err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(http.MethodPut, uploadURL, f)
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", ct)
+
+	resp, err := http.DefaultClient.Do(req.WithContext(ctx))
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("failed to upload image: %s", resp.Status)
+	}
+	return nil
+}
+
+func ensureValidFilename(filename, absDir string) (string, error) {
+	if strings.HasPrefix(filename, "https://") || strings.HasPrefix(filename, "http://") {
+		return "", nil // skip
+	}
+
+	if strings.HasPrefix(filename, "file://") {
+		u, err := url.Parse(filename)
+		if err != nil {
+			return "", err
+		}
+		if u.Host != "" && u.Host != "localhost" {
+			return "", fmt.Errorf("invalid file URL %s", filename)
+		}
+		p := u.Path
+		if strings.HasPrefix(p, "/") && os.PathSeparator == '\\' {
+			p = strings.TrimPrefix(p, "/")
+		}
+		filename = strings.ReplaceAll(p, "/", string(os.PathSeparator))
+		return filename, nil
+	}
+
+	if filepath.IsAbs(filename) {
+		return filename, nil
+	}
+
+	return filepath.Join(absDir, filename), nil
+}
+
+func sha1sum(filename string) (string, error) {
+	f, err := os.Open(filename)
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+
+	s := sha1.New()
+	_, err = io.Copy(s, f)
+	if err != nil {
+		return "", err
+	}
+	result := s.Sum(nil)
+	return fmt.Sprintf("%x", result), nil
+}
+
+type imageFinder struct {
+	images map[string][]imageReference
+	docDir string
+	err    error
+}
+
+func (f *imageFinder) Transform(node *ast.Document, reader text.Reader, pc parser.Context) {
+	if f.images == nil {
+		f.images = make(map[string][]imageReference)
+	}
+	if f.err != nil {
+		return
+	}
+	src := reader.Source()
+	f.err = ast.Walk(node, func(n ast.Node, entering bool) (ast.WalkStatus, error) {
+		if !entering {
+			return ast.WalkContinue, nil
+		}
+
+		var (
+			imgRef           string
+			literalHTML      []byte
+			startPos, endPos int
+		)
+		switch el := n.(type) {
+		case *ast.Image:
+			imgRef = string(el.Destination)
+			p := el.BaseNode.Parent()
+			for p != nil {
+				if p.Kind() == ast.KindParagraph {
+					for i := 0; i < p.Lines().Len(); i++ {
+						// we can have multiple lines in a paragraph, so we need to check each one to match destination/title, hoping there are no dupes
```

Comment:
you probably saw this question coming, but what happens if there are dups? 😄

Reply:
We'll replace the first one we find and leave the other(s).

Reply:
Fixed this case in https://github.com/cloudquery/cloudquery/pull/16033/commits/143410ac2835bc377df7719ae5857ff65827f1bf